src/card_operations/card_fetch_challenge.c

```diff
@@ -0,0 +1,182 @@
+ *****************************************************************************/
+
+/*****************************************************************************
+ * PRIVATE TYPEDEFS
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTION PROTOTYPES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+card_error_type_e card_fetch_challenge(const uint8_t *wallet_name) {
+  // X-Coordinate cannot be more than the number of shares
+  if (NULL == wallet_name) {
+    return CARD_OPERATION_DEFAULT_INVALID;
+  }
+
+  card_error_type_e result = CARD_OPERATION_DEFAULT_INVALID;
+  uint8_t target[SHA256_SIZE] = {0};
+  uint8_t random_number[POW_RAND_NUMBER_SIZE] = {0};
+
+  card_operation_data_t card_data = {0};
+  card_data.nfc_data.retries = 5;
+  card_data.nfc_data.init_session_keys = true;
+
+  char heading[50] = "";
+
+  uint8_t wallet_index;
+  get_index_by_name((const char *)wallet_name, &wallet_index);
+
+  snprintf(heading,
+           sizeof(heading),
+           UI_TEXT_TAP_CARD,
+           decode_card_number(get_wallet_card_locked(wallet_index)));
+
+  while (1) {
+    card_data.nfc_data.acceptable_cards = get_wallet_card_locked(wallet_index);
+    memcpy(card_data.nfc_data.family_id, get_family_id(), FAMILY_ID_SIZE);
+    card_data.nfc_data.tapped_card = 0;
+
+    result = card_initialize_applet(&card_data);
+
+    if (CARD_OPERATION_SUCCESS == card_data.error_type) {
+      card_data.nfc_data.status =
+          nfc_get_challenge(wallet_name, target, random_number);
+
+      if (SW_NO_ERROR == card_data.nfc_data.status) {
+        ASSERT(SUCCESS == add_challenge_flash((const char *)wallet_name,
+                                              target,
+                                              random_number));
+        log_hex_array("pow-rn: ", random_number, POW_RAND_NUMBER_SIZE);
+        LOG_CRITICAL("pow-tg n: %d", pow_count_set_bits(target));
+        buzzer_start(BUZZER_DURATION);
+        break;
+      } else if (SW_WARNING_STATE_UNCHANGED == card_data.nfc_data.status) {
+        // If wallet is not in locked state, then update flash wallet data.
```

Comment:
`If` in the comment changes the meaning of the comment.
```suggestion
        // wallet is in unlocked state; update flash wallet data.
```

Reply:
updated